### PR TITLE
support Conv BN folding for inline inbuilt modules

### PR DESCRIPTION
**Pull request by jiayisunx**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131888

Fix https://github.com/pytorch/pytorch/issues/131693, https://github.com/pytorch/pytorch/issues/130023.
This PR supports Conv BN folding for inline inbuilt modules.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_torchbench_amp_freezing_inference.csv [hunk 1]
```diff
@@ -106,7 +106,7 @@ doctr_det_predictor,pass,5
 
 
 
-doctr_reco_predictor,pass,4
+doctr_reco_predictor,fail_accuracy,4
```
Comment (leslie-fang-intel) [inline benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_torchbench_amp_freezing_inference.csv:109]: Why this model also fail the accuracy? I thought it was `lcnet_50`, isn't it?

File: torch/_inductor/fx_utils.py [hunk 2]
```diff
@@ -52,6 +52,32 @@ def matches_module_function_pattern(
...
+        node, torch.fx.Node
+    ):
+        return False
+    # the second node is call_function or call_method
+    if node.op != "call_function" and node.op != "call_method":
+        return False
+    if node.args[0].target != pattern[0]:
+        return False
+    # the second node is call_function or call_method
+    if node.op != "call_function" and node.op != "call_method":
```
Comment (leslie-fang-intel) [inline torch/_inductor/fx_utils.py:71]: Why check it again?

**Review by leslie-fang-intel — COMMENTED** [review]
Adding a testcase for it? And please also check why previous conv-bn test case didn't catch this failure.

File: torch/_inductor/fx_passes/pre_grad.py [hunk 3]
```diff
@@ -504,6 +509,73 @@ def _used_by_same_conv_module(users):
                     bn_node.replace_all_uses_with(bn_node.args[0])
                     gm.graph.erase_node(bn_node)
     gm.graph.lint()
```
Comment (jgong5) [inline torch/_inductor/fx_passes/pre_grad.py:511]: Remove this?

File: torch/_inductor/fx_utils.py [hunk 4]
```diff
@@ -52,6 +52,32 @@ def matches_module_function_pattern(
     return True
 
 
+def matches_function_pattern(
```
Comment (jgong5) [inline torch/_inductor/fx_utils.py:55]: Add a note what this function does?

**Comment by jiayisunx** [comment]
> Adding a testcase for it? And please also check why previous conv-bn test case didn't catch this failure.

Previous conv-bn test cases didn't cover inline inbuilt modules, and I have added it. Thanks!

File: benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_torchbench_amp_freezing_inference.csv (hunk 1 above)
Comment (jiayisunx) [reply benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_torchbench_amp_freezing_inference.csv:109]: doctr_reco_predictor also in inline_inbuilt_nn_modules_models list, same issue as lcnet_50.

File: torch/_inductor/fx_utils.py (hunk 2 above)
Comment (jiayisunx) [reply torch/_inductor/fx_utils.py:73]: I have corrected it, thanks!

File: torch/_inductor/fx_passes/pre_grad.py (hunk 3 above)
Comment (jiayisunx) [reply torch/_inductor/fx_passes/pre_grad.py]: Done, thanks!

File: torch/_inductor/fx_utils.py (hunk 4 above)
Comment (jiayisunx) [reply torch/_inductor/fx_utils.py:57]: Added, thanks!

File: benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_torchbench_amp_freezing_inference.csv (hunk 1 above)
Comment (leslie-fang-intel) [reply benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_torchbench_amp_freezing_inference.csv:109]: Why lcnet_50 is not affected by this PR?

Comment (jiayisunx) [reply benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_torchbench_amp_freezing_inference.csv:109]: lcnet_50 also is affected by this PR.

**Comment by jiayisunx** [comment]
@leslie-fang-intel , @jgong5 , could you please review this PR again? thanks!

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/fx_passes/pre_grad.py [hunk 5]
```diff
@@ -373,6 +378,7 @@ def disable_fusion(self):
         def is_fusion_enabled(self):
             return self.fusion_enabled
 
+    counters["inductor"]["conv_bn_folding"] = 0
```
Comment (leslie-fang-intel) [inline torch/_inductor/fx_passes/pre_grad.py:381]: Should we reset the count here?

File: torch/_inductor/fx_passes/pre_grad.py [hunk 6]
```diff
@@ -500,9 +506,127 @@ def _used_by_same_conv_module(users):
...
+                if bn_running_mean is None or bn_running_var is None:
+                    continue
+
+                bn_nodes_to_fuse.append(node)
+
+        updated_conv_weights: Dict[str, Any] = {}
+        for node in bn_nodes_to_fuse:
+            counters["inductor"]["conv_bn_folding"] += 1
+            conv_node = node.args[0]
+            if conv_node.args[1].target in updated_conv_weights:
```
Comment (leslie-fang-intel) [inline torch/_inductor/fx_passes/pre_grad.py:584]: nit: also add the comments here.

**Review by leslie-fang-intel — APPROVED** [review]

File: benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_timm_amp_freezing_inference.csv [hunk 7]
```diff
@@ -110,7 +110,7 @@ jx_nest_base,pass,0
 
 
 
-lcnet_050,pass,0
+lcnet_050,fail_accuracy,0
```
Comment (jansel) [inline benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_timm_amp_freezing_inference.csv:113]: Why?

**Review by jansel — CHANGES_REQUESTED** [review]
Could we do this pass as post_grad using the existing pattern matcher (see sdpa patterns) rather than pre-grad?

File: torch/_inductor/fx_passes/pre_grad.py (hunk 5 above)
Comment (jiayisunx) [reply torch/_inductor/fx_passes/pre_grad.py]: I removed this line, thanks!

File: torch/_inductor/fx_passes/pre_grad.py [hunk 8]
```diff
@@ -500,9 +506,127 @@ def _used_by_same_conv_module(users):
...
+                    with gm.graph.inserting_before(conv_node):
+                        conv_bias_name = conv_node.args[1].target.replace(
+                            "weight", "bias"
+                        )
+                        setattr(gm, conv_bias_name, new_conv_bias)
+                        conv_bias_node = gm.graph.create_node(
+                            "get_attr", conv_bias_name, (), {}
+                        )
+                    conv_node.update_arg(2, conv_bias_node)
+                updated_conv_weights[conv_node.args[1].target] = conv_node.args[2]
```
[the remark this answers is not in the document]
Comment (jiayisunx) [reply torch/_inductor/fx_passes/pre_grad.py:630]: Added, thanks!

File: torch/_inductor/fx_passes/pre_grad.py (hunk 6 above)
Comment (jiayisunx) [reply torch/_inductor/fx_passes/pre_grad.py:584]: Done.

**Comment by jiayisunx** [comment]
> Could we do this pass as post_grad using the existing pattern matcher (see sdpa patterns) rather than pre-grad?

If we do this pass as post_grad we will encounter some issues: If we do this pass after `constant_fold`, there will be accuracy issues in the bf16/fp16 cases due to conv weights have been converted to bf16/fp16 data type, if we do this pass before `constant_fold`, it will be complicated, there are various folding patterns because the conv weights may be plugged into various `to` operations(to bf16/fp16 or to CL).

File: benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_timm_amp_freezing_inference.csv (hunk 7 above)
Comment (jiayisunx) [reply benchmarks/dynamo/ci_expected_accuracy/cpu_inductor_timm_amp_freezing_inference.csv:113]: lcnet_050 and doctr_reco_predictor will fail accuracy when conv_bn folding is done. It's a known issue, the [PR](https://github.com/pytorch/pytorch/pull/129775/files) make them pass by adding them in `inline_inbuilt_nn_modules_models `list, this PR supports Conv BN folding for inline inbuilt modules, so the issue is back.

**Comment by jansel** [comment]
Post grad passes are safer because the run after functionalization.   There can be hard to debug correctness issues when you transform a graph with mutation.

**Review by eellison — COMMENTED** [review]
+1, can you extend the existing logic that we have in https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/binary_folding.py instead of doing this in pre-grad

**Comment by jiayisunx** [comment]
> +1, can you extend the existing logic that we have in https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/binary_folding.py instead of doing this in pre-grad

If we do this in `binary_folding`, there will be accuracy issues in bf16/fp16 cases due to conv weights have been converted to bf16/fp16 data type.

**Comment by eellison** [comment]
See PR here. https://github.com/pytorch/pytorch/pull/106576#issue-1835719566

> We fold the weights in fp32 because it's slightly better accuracy, then at the end of the pass convert back the weights to their original dtype.

**Comment by leslie-fang-intel** [comment]
> We fold the weights in fp32 because it's slightly better accuracy, then at the end of the pass convert back the weights to their original dtype.

Hi @eellison, Thanks for the comment. But there still may have the numerical difference, is my understanding right?

- In the pre-grad, the weight of conv is FP32.
- In the post-grad, after constant folding, the weight has been converted to BF16 (here we may lost some accuracy). Then in the binary folding, we convert the BF16 weight back to fp32 implicitly as binary multiply a FP32 frozen parameter. 

Do we know how much it may impact to E2E models?

**Review by jansel — CHANGES_REQUESTED** [review]
Let's move this to join_graph/post_grad

**Comment by leslie-fang-intel** [comment]
Sure, @jansel @eellison we are extending current binary folding implementation and test the models' performance/accuracy. Will keep you updated.

**Pull request closed without merging**